### DISCUSSION TITLE
Ability to auto-activate the first matching option in Combobox

### DISCUSSION THREAD

**Discussion by callumlocke** [tailwindlabs/headlessui#1320] [Help] [answered]
Hello. I'm hoping there's a way to do this already, otherwise I'd like to propose it as a feature.

Using the [demo in the docs](https://headlessui.dev/react/combobox), if you type "T", you get three matching options, like this:

<img width="322" alt="image" src="https://user-images.githubusercontent.com/250617/162739611-dd241f9a-4e19-4d13-b52d-07b3be21458d.png">

...but if you press "Enter" at this point, it just changes the box back to the default option ("Wade Cooper"). In my particular use case, this feels weird and unexpected. To get it to actually choose, say, the first option ("Tom Cook"), I have to press "Down" first before pressing "Enter".

What I'd like: a way to force it so that, whenever options are visible, there is **always** an option activated (the first, by default), i.e. so it would look like this instead:

<img width="337" alt="image" src="https://user-images.githubusercontent.com/250617/162740561-0be03bdf-7d4b-4533-b031-507e585c8c58.png">

This way, I can just type a few characters, and if the first option is the one I want, I can just hit "Enter" to select it, without having to press the "Down" arrow key first. Is there any way to achieve this behaviour with the existing API?

**Comment by callumlocke** [comment]
I attempted a hack: I passed a `ref` to `ComboBox.Input` (this seems to be passed to the actual `input` element correctly), and then in a `useEffect` I dispatched a fake `new KeyboardEvent('keydown', { key: 'ArrowDown' })` on the `input` (whenever there's a query entered *and* the `activeIndex` is reported by `ComboBox` to be `null` *and* the input is focused)... but for some reason this doesn't work. Looking at the source, I feel like it should, but it doesn't. Any suggestions welcome.

**Comment by HemalR** [comment]
Agree with the UI flow you have laid out. I think we could figure out a hacky way to allow it, but it is unlikely to be ideal and would have flaws. This seems like a good default, if not, I would like to see it as an option via a prop.

**Comment by jdlien** [comment]
@callumlocke Something like this could be an improvement. The way the input works now, users can type in something and have the impression that they have selected something when they have not. Without some kind of rigorous front-end validation, it can lead to a confusing UX because the option won't have changed when a user may think it has.

**Comment by RobinMalfait** [comment, marked as answer]
Hey! Thank you for your suggestion!
Much appreciated! 🙏

This should be fixed, and will be available in the next release.

You can already try it using:
- `npm install @headlessui/react@insiders` or `yarn add @headlessui/react@insiders`.
- `npm install @headlessui/vue@insiders` or `yarn add @headlessui/vue@insiders`.

**Comment by Seolhun** [comment]
Very interesting.
I had the problem that the activeIndex starts from 0 when I press Enter on the other hand.

I made an Autocomplete Component with Combobox. When the User hits Enter, I don't know if the user wants to enter a search or select the index item of 0. Therefore, I need a way to tell this apart. If activeIndex starts from 0, this problem cannot be solved.

It would be nice to have props that can control the activeIndex. Do you have good ideas fir this? 🙏🏽

**Reply by HemalR** [reply]
What if your index option `0` was something along the lines of: `Search for ${SEARCH_QUERY}`

Editing the exact text as per your use case, but essentially, making the first option a dynamic option that, when selected, does the 'search' you are referring to.

**Reply by zipme** [reply]
I have the same issue, using Combobox for the global search input field, what I want to achieve is if the user hit enter, it redirects to the search result page instead of selecting the first option. Is there a way to disable this?

**Reply by afrikazil** [reply]
absolutely the same

**Reply by DogAndHerDude** [reply]
Absolutely in the same boat. This really messes with the input by basically just being a searchable ListBox. You really cannot have any control on the values.

Why not just have an extra prop `autoFocusOption` or something that would focus the first matching option if `true`. Otherwise it sets the `activeIndex` to `-1` allowing us to have more control on custom values instead of having to forcefully select the suggested value even if the user does not want to.

**Reply by deanmikan** [reply]
Agree with above, allow us to disable auto-selecting the first option!

**Reply by mhafren** [reply]
I would also like to disable to auto-select. Otherwise working fine!

**Reply by amazingmaze** [reply]
Same here. Works fine but this auto-select function breaks the component for use as a "regular" search.

**Reply by lcnogueira** [reply]
@zipme , were you able to find a solution for what you mentioned here https://github.com/tailwindlabs/headlessui/discussions/1320#discussioncomment-5641880 ?

**Reply by Fabioni** [reply]
same here: auto-select function breaks the component for use as a "regular" search

**Reply by Rich5** [reply]
Same here. Is there anyway to disable this behavior?

**Reply by ixgn** [reply]
Bump, same here @RobinMalfait
